aptos-move/framework/aptos-framework/sources/staking_contract.move

```diff
@@ -0,0 +1,1132 @@
+    fun distribute_internal(
+        staker: address,
+        operator: address,
+        staking_contract: &mut StakingContract,
+        distribute_events: &mut EventHandle<DistributeEvent>,
+    ) {
+        let pool_address = staking_contract.pool_address;
+        let (_, inactive, _, pending_inactive) = stake::get_stake(pool_address);
+        let total_potential_withdrawable = inactive + pending_inactive;
+        let coins = stake::withdraw_with_cap(&staking_contract.owner_cap, total_potential_withdrawable);
+        let distribution_amount =  coin::value(&coins);
+        if (distribution_amount == 0) {
+            coin::destroy_zero(coins);
+            return
+        };
+
+        let distribution_pool = &mut staking_contract.distribution_pool;
+        update_distribution_pool(
+            distribution_pool, distribution_amount, operator, staking_contract.commission_percentage);
+
+        // Buy all recipients out of the distribution pool.
+        while (pool_u64::shareholders_count(distribution_pool) > 0) {
+            let recipients = pool_u64::shareholders(distribution_pool);
+            let recipient = *vector::borrow(&mut recipients, 0);
+            let current_shares = pool_u64::shares(distribution_pool, recipient);
+            let amount_to_distribute = pool_u64::redeem_shares(distribution_pool, recipient, current_shares);
+            coin::deposit(recipient, coin::extract(&mut coins, amount_to_distribute));
+
+            emit_event(
+                distribute_events,
+                DistributeEvent { operator, pool_address, recipient, amount: amount_to_distribute }
+            );
+        };
+
+        // In case there's any dust left, send them all to the staker.
+        if (coin::value(&coins) > 0) {
+            coin::deposit(staker, coins);
+        } else {
+            coin::destroy_zero(coins);
+        }
+    }
+
+    // Assert that a staking_contract exists for the staker/operator pair.
+    fun assert_staking_contract_exists(staker: address, operator: address) acquires Store {
+        assert!(exists<Store>(staker), error::not_found(ENO_STAKING_CONTRACT_FOUND_FOR_STAKER));
+        let staking_contracts = &mut borrow_global_mut<Store>(staker).staking_contracts;
+        assert!(
+            simple_map::contains_key(staking_contracts, &operator),
+            error::not_found(ENO_STAKING_CONTRACT_FOUND_FOR_OPERATOR),
+        );
+    }
+
+    // Add a new distribution for `recipient` and `amount` to the staking contract's distributions list.
+    fun add_distribution(
+        operator: address,
+        staking_contract: &mut StakingContract,
+        recipient: address,
+        coins_amount: u64,
+        add_distribution_events: &mut EventHandle<AddDistributionEvent>,
+    ) {
+        let distribution_pool = &mut staking_contract.distribution_pool;
+        let (_, _, _, total_distribution_amount) = stake::get_stake(staking_contract.pool_address);
+        update_distribution_pool(
```

Comment:
why do we need this? from what I read, it can only happen after `distribute_internal` (pool being empty). it's called in `request_commission_internal` and `unlock_stake`, both are always precedented by a `distribute_internal`.

some invariants like when the pool is guaranteed to be empty would help reviews.

Reply:
distribute_internal only distributes out the inactive stake, and pending_inactive still stays (with accumulated rewards). The distribution pool is for pending_inactive stake.